contracts/AirDrop.sol

```diff
@@ -0,0 +1,186 @@
+        require(signature.length == 64, "Invalid signature length");
+
+        // assemble input data
+        bytes memory input = new bytes(129);
+        _bytesConcat(input, pubKey, 0, 33);
+        _bytesConcat(input, signature, 33, 64);
+        _bytesConcat(input, _bytes32toBytes(messageHash), 97, 32);
+
+
+        bytes memory output = new bytes(20);
+        /* solium-disable-next-line */
+        assembly {
+          // call tmSignatureRecover precompile contract
+          // Contract address: 0x69
+          let len := mload(input)
+          if iszero(staticcall(not(0), 0x69, input, len, output, 20)) {
+            revert(0, 0)
+          }
+        }
+        
+        // return the recovered address
+        return output;
+    }
+
+    function _bytesConcat(bytes memory data, bytes memory _bytes, uint256 index, uint256 len) internal pure {
+        for (uint i; i<len; ++i) {
+          data[index++] = _bytes[i];
+        }
+    }
+
+    function _bytes32toBytes(bytes32 _data) public pure returns (bytes memory) {
+        return abi.encodePacked(_data);
+    }
+
+    function _tmSignatureHash(
+        bytes32 tokenSymbol,
+        uint256 amount,
+        address recipient
+    ) internal pure returns (bytes32) {
+        return sha256(abi.encodePacked(
+            '{"account_number":"0","chain_id":"',
+            sourceChainID,
+            '","data":null,"memo":"","msgs":[{"amount":"',
+            _bytesToHex(abi.encodePacked(amount), false),
+            '","recipient":"',
+            _bytesToHex(abi.encodePacked(recipient), true),
+            '","token_symbol":"',
+            _bytesToHex(abi.encodePacked(tokenSymbol), false),
+            '"}],"sequence":"0","source":"0"}'
+        ));
+    }
+
+    function _bytesToHex(bytes memory buffer, bool prefix) public pure returns (string memory) {
+        // Fixed buffer size for hexadecimal conversion
+        bytes memory converted = new bytes(buffer.length * 2);
+
+        bytes memory _base = "0123456789abcdef";
+
+        for (uint256 i = 0; i < buffer.length; i++) {
+            converted[i * 2] = _base[uint8(buffer[i]) / _base.length];
+            converted[i * 2 + 1] = _base[uint8(buffer[i]) % _base.length];
+        }
+
+        if (prefix) {
+            return string(abi.encodePacked('0x',converted));
+        }
+        return string(converted);
+    }
+
+    /*********************** Param update ********************************/
+    function updateParam(string calldata key, bytes calldata value) external override onlyInit onlyGov{
+      if (Memory.compareStrings(key,"approvalAddress")) {
+        require(value.length == 20, "length of approvalAddress mismatch");
+        address newApprovalAddress = BytesToTypes.bytesToAddress(20, value);
+        require(newApprovalAddress != address(0), "approvalAddress should not be zero");
+        approvalAddress = newApprovalAddress;
+      } else if (Memory.compareStrings(key,"merkleRoot")) {
```

Comment:
Can we not let `merkleRoot` change in day1? 

Because the consequence is not predictable

Reply:
It's only can be set by gov once  now